### PR TITLE
Module tabulation

### PR DESCRIPTION
**Pull request by cgarciae**
# What does this PR do?
Adds mechanism to tabulate a Module, similar to Haiku's `tabulate` or Kera's `summary`.  This builds on the ideas/comments from #288 and aims to use `rich` to ultimately generate a nice table representation of the module. As previously discussed, the idea is to implement 2 steps:

1. Generate an intermediate representation
2. Render such representation

The PR has the following goals:

- [x] Create function to extract an intermediate representation for the Module (e.g. `module_info`).
- [x] Create a function that takes the intermediate representation and render is as a string (e.g. `render_table`).
- [x] Add a method to Module that directly render the module to a string (e.g. `Module.tabulate`).
- [x] Add docs
- [x] Add tests

**NOTE**: current names used are open for debate.

### CODE REVIEW

**Comment by cgarciae** [comment]
@avital @marcvanzee 

The proposed `module_info` function takes in a Module and sample inputs and returns a `Table` which is just a list of `Row`s, where Row is:

```python
@dataclass
class Row:
  path: Tuple[str, ...]
  outputs: Any # outputs of the module, shapes and dtypes only
  module_variables: Dict[str, Dict[str, Array]] # module's immediate variables only 
  counted_variables: Dict[str, Dict[str, Array]] # all variables that should be counted for this row
```
The idea is that with this information you can later create a meaningful representation. Under the hood its implemented using `jax.eval_shape` over `Module.init` with `capture_intermediate=True`. It has a `depth` argument which can be used to abbreviate the table e.g. `depth=1` will stop the recursion after 1 level so it will only show the top Module and its immediate submodules. `counted_variables` specifies which variables should be counted for that row (e.g. total number or total bytes), in most cases it will be the same as `module_variables` except for when max depth is reached in which case `counted_variables` contains the variables for that module and all of its sub-modules.

**Comment by cgarciae** [comment]
Next I'll implement a `render_table` function that uses rich to create a nice table representation. Design-wise it will look somewhat like this but without the `module` and `inputs` columns:
150032598-3e39f1a4-1a88-4303-9a95-6a8784a8b894

The last columns are going to be variable depending on the collections present e.g `Rng` will be removed and `Parameter` and `BatchStat` will be replace with `params` and `batch_stats`.

**Comment by cgarciae** [comment]
Gave it a go at an initial implementation of `render_table`, here is a sample of how it currently looks:

Screenshot from 2022-02-02 09-18-38

Some notes:
* `variables` and `outputs` are formatted with `yaml`.
* arrays are represented as `{dtype}[{shape}]`, similar how JAX represents tracers.
* need to add checks to avoid empty collections being shown

Feedback on design appreciated.

**Comment by cgarciae** [comment]
@avital 

Even though `ConvBlock` is defined like this:

```python
@dataclasses.dataclass
class ConvBlock(nn.Module):
  features: int
  kernel_size: List[int]
  training: bool

  @nn.compact
  def __call__(self, x: Array) -> Array:
      x = nn.Conv(self.features, self.kernel_size)(x)
      x = nn.BatchNorm(use_running_average=not self.training)(x)
      x = nn.Dropout(0.5, deterministic=not self.training)(x)
      x = nn.relu(x)
      return x
```
Since the `variables` keys are alphabetically ordered e.g:

```
    intermediates: {
        ConvBlock_0: {
            BatchNorm_0: {
                __call__: (ShapeDtypeStruct(shape=(32, 28, 28, 32), dtype=float32),),
            },
            Conv_0: {
                __call__: (ShapeDtypeStruct(shape=(32, 28, 28, 32), dtype=float32),),
            },
            Dropout_0: {
                __call__: (ShapeDtypeStruct(shape=(32, 28, 28, 32), dtype=float32),),
            },
            __call__: (ShapeDtypeStruct(shape=(32, 28, 28, 32), dtype=float32),),
        },
        ConvBlock_1: {
            BatchNorm_0: {
                __call__: (ShapeDtypeStruct(shape=(32, 28, 28, 64), dtype=float32),),
            },
            Conv_0: {
                __call__: (ShapeDtypeStruct(shape=(32, 28, 28, 64), dtype=float32),),
            },
            Dropout_0: {
                __call__: (ShapeDtypeStruct(shape=(32, 28, 28, 64), dtype=float32),),
            },
            __call__: (ShapeDtypeStruct(shape=(32, 28, 28, 64), dtype=float32),),
        },
        Dense_0: {
            __call__: (ShapeDtypeStruct(shape=(32, 10), dtype=float32),),
        },
        __call__: ((ShapeDtypeStruct(shape=(32, 10), dtype=float32), {'a': ShapeDtypeStruct(shape=(32, 10), dtype=float32), 'b': ShapeDtypeStruct(shape=(32, 10), dtype=float32)}),),
```
It would appear as if `BatchNorm` happened before `Conv` which can be confusing if you don't know about this.

**Comment by cgarciae** [comment]
I thought that it would be nice if we showed the shapes of the inputs to the main module so I added a row with that. Here is how the latest version looks:

Screenshot from 2022-02-03 08-31-48
 
 Here is the same Module but simplified with `depth=1`:
 
 
Screenshot from 2022-02-03 08-32-18

**Comment by avital** [comment]
This looks really cool! One question about the design -- why make separate columns for each variable collection? Seems potentially wasteful as these columns are always only one line, whereas the others may be multiple lines. Instead, maybe you could fit the parameter count and memory footprint directly within the column for variables, something like this maybe?

[...oh! you did the layout this way so that you could summarize the total at the bottom...]

So, maybe another proposal is to drop the `variables` column entirely and embed all of the information that's there into the respective column for each of the variable collections, e.g.:

```
batch_stats
---

64 (256 B)
  mean: float32[32]
  var: float32[32]
```

**Comment by cgarciae** [comment]
@avital we could strive for removing the `variables` column, however, it might look strange when you use the `depth` parameter to get less rows. Currently the `variables` column shows the variables created by that module only (not the variables by its submodules) and columns like `params` or `batch_stats` shows the counts for that module and all its submodules when its submodules are filtered out by `depth`.

This is under the assumption we want to have the `depth` filter / summarization mechanism (do we?). I kinda like `depth` because when for big models it can give you a better overview if you set depth to low values like `1` or `2`. That said, this is very subjective an I am open to ideas.

**Comment by cgarciae** [comment]
@avital I implemented your suggestion of removing the `variables` column and I quick like it:

Screenshot from 2022-02-04 10-06-47

Here is with `depth=1`:

Screenshot from 2022-02-04 10-07-44

**Comment by avital** [comment]
Hi @cgarciae and sorry for the late reply. I agree, I think this looks very nice!

**Comment by cgarciae** [comment]
@avital I proceeded to create the `tabulate` method. Will be adding tests next.

Also, this PR introduced a lot of helper function and some classes, organization-wise are they ok there or should we move them?

**Comment by avital** [comment]
@cgarciae I propose moving all of the new code to flax/linen/summary.py rather than module.py. I guess we should keep the `tabulate` method on Module for sake of discoverability but all of its implementation is probably best in a separate file. And I propose treating `module_info` (maybe rename `module_summary` in line with my proposed file name? Or we could unify on some different name) and `render_table` as public APIs as people may want to, e.g. run `module_info` on one machine and `render_table` on another, e.g. when serializing...

**Comment by cgarciae** [comment]
@avital thanks for the feedback! 

* Moved helper functions/classes to `summary.py` as suggested
* Rename `module_info` to `module_summary`. 
* Exported `module_summary` and `render_table` as public attributes in the `linen` module.
* Added 3 tests.
* Added docs for `module_summary` and `render_table`.

**Comment by cgarciae** [comment]
Hey @jheek! I had a talk with @marcvanzee, he suggested that I asked you if you could review this PR since both him and @avital are unavailable right now 🙏‍

**Comment by marcvanzee** [comment]
@cgarciae apologies for the superlate reply. As you know we have been very busy and understaffed lately, but hopefully that will become better now! I will review your PR now. Can you please rebase and resolve the conflicts in the meantime?

**Comment by cgarciae** [comment]
Hey @marcvanzee! Fixed merge issues.

Check the `test_tabulate` test if you want to play with it. One last pending issue to discuss here is that rows are not ordered by call order but alphabetically instead, in the future it would nice to modify the `capture_intermediates` feature to somehow indicate call order as well.

cc: @jheek

**Review by marcvanzee — CHANGES_REQUESTED** [review]
First of all, this is going to be a really great addition! Users often struggle with printing their params and they always use their own little util functions.

As a high-level comment: I am having some problems rendering this table in the Python interpreter:

```
>>> nn.Dense(4).tabulate(random.PRNGKey(0), jnp.ones((3, 4)))
'\n\x1b[3m                 Dense Summary                  \x1b[0m\n┏━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━┓\n┃\x1b[1m \x1b[0m\x1b[1mpath  \x1b[0m\x1b[1m \x1b[0m┃\x1b[1m \x1b[0m\x1b[1moutputs     \x1b[0m\x1b[1m \x1b[0m┃\x1b[1m \x1b[0m\x1b[1mparams              \x1b[0m\x1b[1m \x1b[0m┃\n┡━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━┩\n│ Inputs │ \x1b[2mfloat32\x1b[0m[3,4] │                      │\n├────────┼──────────────┼──────────────────────┤\n│ *      │ \x1b[2mfloat32\x1b[0m[3,4] │ bias: \x1b[2mfloat32\x1b[0m[4]     │\n│        │              │ kernel: \x1b[2mfloat32\x1b[0m[4,4] │\n│        │              │                      │\n│        │              │ \x1b[1m20 \x1b[0m\x1b[1;2m(80 B)\x1b[0m            │\n├────────┼──────────────┼──────────────────────┤\n│\x1b[1m \x1b[0m\x1b[1m      \x1b[0m\x1b[1m \x1b[0m│\x1b[1m \x1b[0m\x1b[1m       Total\x1b[0m\x1b[1m \x1b[0m│\x1b[1m \x1b[0m\x1b[1m20 \x1b[0m\x1b[1;2m(80 B)\x1b[0m\x1b[1m           \x1b[0m\x1b[1m \x1b[0m│\n└────────┴──────────────┴──────────────────────┘\n\x1b[1m                                                \x1b[0m\n\x1b[1m          Total Parameters: 20 \x1b[0m\x1b[1;2m(80 B)\x1b[0m\x1b[1m           \x1b[0m\n\n'
```

This happens both in my ITerm terminal and in my VSCode terminal.

I am using UTF-8 encoding and I have coloring enabled in Iterm. Any idea why this happens?

I suppose something is wrong with my setup, but given that I run into this now, I fear that it may happen to other users as well...

File: flax/linen/summary.py
Comment (marcvanzee) [inline flax/linen/summary.py:59]: The name seems to suggests this will return a summary, but it actually returns a Table object, which isn't very useful, so I wonder if it is useful to expose this function at all. 

Alternatively you could add a function `tabulate` to this file that calls `module_summary` (I would rename it to `_get_module_table`) and `render_table` (I would rename it to `_render_table`).

Otherwise, users may try to call `summary.module_summary` expecting a summary of the model.

**Review by marcvanzee — CHANGES_REQUESTED** [review]

File: flax/linen/summary.py
Comment (cgarciae) [reply flax/linen/summary.py:91]: This is shorter but I am not sure its clearer 🤔

File: flax/linen/module.py
Comment (cgarciae) [reply flax/linen/module.py:1363]: Sounds good, I was aiming for `tabulate` to have (almost) the same signature as `init` so this was copied over, but it wouldn't hurt to have `tabulate` be as flexible as possible.

File: flax/linen/summary.py
Comment (cgarciae) [reply flax/linen/summary.py:59]: I like this! Will do.

BTW: I wish `Module.init/apply` did this as well.

File: flax/linen/__init__.py
Comment (cgarciae) [reply flax/linen/__init__.py]: Doesn't register as a change from this PR 🤔

File: flax/linen/summary.py
Comment (cgarciae) [reply flax/linen/summary.py:91]: Ok, will use the short version.

File: flax/linen/summary.py
Comment (cgarciae) [reply flax/linen/summary.py:59]: Plan:
- rename `module_summary` => `_get_module_table`
- rename `render_table` => `_render_table`
- create `summary.tabulate` that follows @jheek transformation style suggestion.

File: flax/linen/summary.py
Comment (cgarciae) [reply flax/linen/summary.py:169]: I'll rename it to `path_repr` which makes more sense.

Comment (cgarciae) [reply flax/linen/summary.py:169]: On second thoughts I just inlined it.

File: tests/linen/linen_module_test.py
Comment (cgarciae) [reply tests/linen/linen_module_test.py:17]: sounds good!

File: tests/linen/linen_module_test.py
Comment (cgarciae) [reply tests/linen/linen_module_test.py]: Pytest somehow overrides `assert` such that if you change this `11` to a `12` the error message when you run this test like this:

```bash
pytest tests/linen/summary_test.py
```
is presented in a rich form as:
```
>     assert len(table) == 12
E     AssertionError: assert 11 == 12
E      +  where 11 = len([Row(path=('Inputs',), outputs=(DeviceArray([[[[1.],\n               [1.],\n               [1.],\n               ...,\n   ...2), {'a': ShapeDtypeStruct(shape=(32, 10), dtype=float32), 'b': ShapeDtypeStruct(shape=(32, 10), dtype=float32)}),)}})])
```
Using `assertEquals` looks like this:
```
>     self.assertEqual(len(table), 12)
E     AssertionError: 11 != 12
```

Comment (cgarciae) [reply tests/linen/linen_module_test.py]: Ok, I see `assertEqual` is used throughout the code base so I will change it to match the style.

Comment (cgarciae) [reply tests/linen/linen_module_test.py]: I do prefer the `assert` style as it looks more natural.

File: flax/linen/module.py
Comment (cgarciae) [reply flax/linen/module.py:1435]: Added brief description for `*args` and `**kwargs`.

**Comment by cgarciae** [comment]
Hey @marcvanzee and @jheek!
Fixed a couple issues based on your feedback. Its ready for another round 😀

**Comment by jheek** [comment]
@cgarciae you can have at most 5 commits in a PR. Please squash / rebase to fix the build

**Review by jheek — CHANGES_REQUESTED** [review]

File: flax/linen/module.py
Comment (jheek) [inline flax/linen/module.py:1358]: tabulate does not yet show up in the docs. You still need to add it to flax.linen.rst Note you can click "details" on the RTD action to see a preview of how the docs will look

Comment (jheek) [reply flax/linen/module.py:1358]: I think inspection/summarization makes sense

**Comment by marcvanzee** [comment]
> I think inspection/summarization makes sense

+1 to inspection!

**Comment by cgarciae** [comment]
Created the `Inspection` section.

File: flax/linen/summary.py
Comment (marcvanzee) [inline flax/linen/summary.py:389]: Imports should be at the top.

File: flax/linen/summary.py
Comment (marcvanzee) [inline flax/linen/summary.py:300]: Nit: Remove?

File: flax/linen/summary.py
Comment (marcvanzee) [inline flax/linen/summary.py:304]: Nit: Remove?

File: flax/linen/summary.py
Comment (marcvanzee) [inline flax/linen/summary.py:325]: Nit: Remove?

File: flax/linen/summary.py
Comment (marcvanzee) [inline flax/linen/summary.py:335]: typo: `colletion_total` --> `collection_total`

File: flax/linen/summary.py
Comment (marcvanzee) [inline flax/linen/summary.py:359]:
```suggestion
```

File: flax/linen/summary.py
Comment (marcvanzee) [inline flax/linen/summary.py:384]: Note: `bytes` is a reserved keyword in Python.

```suggestion
  leaves = jax.tree_leaves(pytree)
  size = sum(x.size for x in leaves)
  num_bytes = sum(x.size * x.dtype.itemsize for x in leaves)
```

File: flax/linen/summary.py
Comment (marcvanzee) [inline flax/linen/summary.py:399]: Isn't this more general?

```suggestion
  if not value:
```

File: flax/linen/summary.py
Comment (marcvanzee) [inline flax/linen/summary.py:418]: How about this? It will output shapes as `(3, 4)` instead of `[3, 4]`, but that is actually quite standard. Or did you have a reason why you want to use square brackets?

```suggestion
def _format_value(value):
  if isinstance(value, (np.ndarray, jnp.ndarray)):
    return f"[dim]{value.dtype}[/dim][{value.shape}]"
  else:
    return str(value)
```

**Review by marcvanzee — CHANGES_REQUESTED** [review]
Some overall comments:

* The build is failing in this PR, but the error looks unrelated. Maybe just sync to main will resolve it?
* We use single quotes in the Flax codebase, so please change that everywhere.

File: flax/linen/module.py
Comment (cgarciae) [reply flax/linen/module.py:1443]: Sound good.

File: flax/linen/summary.py
Comment (cgarciae) [reply flax/linen/summary.py:35]: Makes sense! I think they where created before the docs strings just so I knew what they did 😅

File: flax/linen/summary.py
Comment (cgarciae) [reply flax/linen/summary.py:333]: good catch!

File: flax/linen/summary.py
Comment (cgarciae) [reply flax/linen/summary.py:382]: Changed to `_bytes`

File: flax/linen/summary.py
Comment (cgarciae) [reply flax/linen/summary.py]: I'll be keeping
```python
if hasattr(value, "shape") and hasattr(value, "dtype"):
```
as `ShapeDtypeStruct` created by `eval_shape` is not detected as an instance of `jnp.ndarray`.

File: flax/linen/summary.py
Comment (cgarciae) [reply flax/linen/summary.py]: Done!

**Comment by cgarciae** [comment]
Hey @marcvanzee! Thanks for detailed review.

I think the only thing missing is how do we want to render arrays:

With this change values will render like this:

**original**
```
bias: float32[2]
kernel: float32[4,2]
```
**new proposal**
```
bias: float32[(2,)]
kernel: float32[(4, 2)]
```
**alternative proposal**
```
bias: float32(2)
kernel: float32(4, 2)
```

File: flax/linen/summary.py
Comment (jheek) [reply flax/linen/summary.py]: A square bracket shape is quite standard in rendering I think. See for example how jaxpr's are rendered when you print the output of jax.make_jaxpr. I think the intuition borrows from the fact that you always index a multi-dimensional array with square brackets but instead of the index you now put the length.

**Comment by marcvanzee** [comment]
> Hey @marcvanzee! Thanks for detailed review.
> 
> I think the only thing missing is how do we want to render arrays:
> 
> **original**
> 
> ```
> bias: float32[2]
> kernel: float32[4,2]
> ```
> 
> **new proposal**
> 
> ```
> bias: float32[(2,)]
> kernel: float32[(4, 2)]
> ```
> 
> **alternative proposal**
> 
> ```
> bias: float32(2)
> kernel: float32(4, 2)
> ```

Having read @jheek , I think the original was actually fine, so feel free to keep using that!

File: flax/linen/summary.py
Comment (cgarciae) [reply flax/linen/summary.py]: If `value` is something like `0` or `None` it would give the wrong output.

**Comment by cgarciae** [comment]
@marcvanzee reverted back to square brackets `[...]`. Only one comment pending for resolution.

File: flax/linen/summary.py
Comment (marcvanzee) [reply flax/linen/summary.py]: Make sense! Let's keep it as is then.

File: tests/linen/linen_module_test.py
Comment (marcvanzee) [reply tests/linen/linen_module_test.py]: `self.assertEqual` provides a bit more functionality than just `assert`. It provides better error messages in the case of failure, and it will automatically call `assertListEqual`, `assertDictEqual`, ... if the two inputs have the same type.

**Review by marcvanzee — APPROVED** [review]
Looks good to me!

Minor comment for the next time: It wasn't entirely clear to me whether I should review the PR again or whether you were still addressing my comments. You can explicitly re-request a review in the main tab of the PR by clicking the cycled arrows next to a reviewer's name. This is quite helpful since it can happen that you are already addressing some of my comments, but you do not yet want me to do the review again. So then it helps if we both know that I won't do another review until you've explicitly requested it.

**Review by jheek — APPROVED** [review]
Looks great!

**Comment by marcvanzee** [comment]
@cgarciae this is currently blocked on internal tests failing. We should just make sure we link in `rich` and get access to it, I'll look into it.

**Comment by marcvanzee** [comment]
@cgarciae this change seems to add back all deprecated HOWTOs. perhaps something went wrong when syncing to main, could you please revert those changes?

**Comment by marcvanzee** [comment]
@cgarciae Update: I will update it in the internal version so no action needed.

**Comment by cgarciae** [comment]
Ok, or should I just sync again with main?

**Comment by cgarciae** [comment]
@marcvanzee I cleaned the branch, don't know why it caught these other changes.

**Comment by marcvanzee** [comment]
This change is committed but the PR is still open: https://github.com/google/flax/commit/ef6bf4054c30271a58bfabb58f3d0049ef5d851a

I am closing this PR, thank you @cgarciae!!! Awesome change!

**Comment by trvs2cool** [comment]
This works really nicely for Modules using the call method.  

However, for modules using other method, it does not work.  The issue is that module.init with capture_intermediates=True only captures [__call__](https://github.com/google/flax/blob/8a148dbd5de4c327dd6b7c931a8b649431e8274e/flax/linen/module.py#L1145) intermediates (See [summary.py:202](https://github.com/google/flax/blob/8a148dbd5de4c327dd6b7c931a8b649431e8274e/flax/linen/summary.py#L202)).

I think the fix is fairly simple.  In [summary.py:184](https://github.com/google/flax/blob/8a148dbd5de4c327dd6b7c931a8b649431e8274e/flax/linen/summary.py#L184)) just pass the output_methods into the capture_intermediates by a lambda:

```  
    capture_intermediates_fn = lambda x,y: y in output_methods or y == '__call__'

    shape_variables = jax.eval_shape(lambda: module.init(
        rngs,
        *args,
        method=method,
        mutable=mutable,
        capture_intermediates=capture_intermediates_fn,
        **kwargs,
    ))```

**Comment by marcvanzee** [comment]
@trvs2cool thanks for spotting this! Do you mind creating a separate issue for this problem?

**Pull request closed without merging**
